page-objects/src/components/sidebar/WelcomeContent.ts

```diff
@@ -0,0 +1,80 @@
+import * as assert from "assert";
+import { By, WebElement } from "selenium-webdriver";
+import { ViewSection } from "./ViewSection";
+import { AbstractElement } from "../AbstractElement";
+
+/**
+ * A button that appears in the welcome content and can be clicked to execute a command.
+ *
+ * To execute the command bound to this button simply run: `await button.click();`.
+ */
+export class WelcomeContentButton extends AbstractElement {
+    /**
+     * @param panel  The panel containing the button in the welcome section
+     * @param welcomeSection  The enclosing welcome section
+     */
+    constructor(panel: WebElement, welcomeSection: WelcomeContentSection) {
+        super(panel, welcomeSection);
+    }
+
+    /** Return the title displayed on this button */
+    public async getTitle(): Promise<string> {
+        return this.getText();
+    }
+}
+
+/**
+ * A section in an empty custom view, see:
+ * https://code.visualstudio.com/api/extension-guides/tree-view#welcome-content
+ *
+ * The welcome section contains two types of elements: text entries and buttons that can be bound to commands.
+ * The text sections can be accessed via [[getTextSections]], the buttons on the
+ * other hand via [[getButtons]].
+ * This however looses the information of the order of the buttons and lines
+ * with respect to each other. This can be remedied by using [[getContents]],
+ * which returns both in the order that they are found (at the expense, that you
+ * now must use typechecks to find out what you got).
+ */
+export class WelcomeContentSection extends AbstractElement {
+    /**
+     * @param panel  The panel containing the welcome content.
+     * @param parent  The webelement in which the welcome content is embedded.
+     */
+    constructor(panel: WebElement, parent: ViewSection) {
+        super(panel, parent);
+    }
+
+    /**
+     * Combination of [[getButtons]] and [[getTextSections]]: returns all entries in the welcome view in the order that they appear.
+     */
+    public async getContents(): Promise<(WelcomeContentButton|string)[]> {
+        const elements = await this.findElements(By.xpath(".//a[@class='monaco-button monaco-text-button'] | .//p"));
+        return Promise.all(elements.map(async (e) => {
+            const tagName = await e.getTagName();
+            if (tagName === "p") {
+                return e.getText();
+            } else if (tagName === "a") {
+                return new WelcomeContentButton(e, this);
+            } else {
+                assert(false, `got an invalid tag name: '${tagName}', expected 'p' or 'a'`);
```

Comment:
wouldn't throwing an error be simpler?

Reply:
Well, kinda. This is more of a "philosophical" question though: my reasoning here is that we expect only buttons and text and anything else means we have a bug in our code (thus we assert that this does not happen). An exception on the other hand means that a user can reach this code path due to an error. But this is something that **must** not happen.

However, that's just my stance on this, so feel free to disagree and I'll change it to throw or remove the else branch entirely.

Reply:
From the API standpoint, `assert(false)` is just a way to throw an `AssertionError`. And I'd say these should be reserved for failing test conditions. The page objects are not really a test framework (or an assertion framework either).

Reply:
I have removed the current else branch entirely and moved the else if into the else branch.